types/codec.go

```diff
@@ -28,13 +28,56 @@ func CanonicalSignBytes(m codec.ProtoMarshaler) ([]byte, error) {
 		return nil, err
 	}
 
-	genericJSON := make(map[string]interface{})
+	var genericJSON interface{}
 
 	// decode canonical proto encoding into a generic map
 	if err := jsonc.Unmarshal(buf.Bytes(), &genericJSON); err != nil {
 		return nil, err
 	}
 
+	// strip default values, i.e. 0, true, "", [], {}, null
+	genericJSONNoDefaults := JSONStripDefaults(genericJSON)
+
 	// finally, return the canonical JSON encoding via JSON Canonical Form
-	return jsonc.Marshal(genericJSON)
+	return jsonc.Marshal(genericJSONNoDefaults)
+}
+
+func JSONStripDefaults(val interface{}) interface{} {
```

Comment:
This won't work for various reasons:

1. The float case is wrong
2. It's missing other primitive types
3. End result still includes `nil` values.

I already have a solution for (1) and (2) but (3) I'm not sure how to deal with yet.

Reply:
I just want to note that I don't think 1 and 2 are necessarily true. Go decodes `0` to a `float64` by default so I think that is the only primitive integer case we need to check for. At least that's what my tests indicate. If you prefer doing the reflection based approach for zero values for safe measure that's fine.

As for 3, I think `null`s are only included in arrays because this current implementation converts default values in arrays to `null`. I actually think that is a bit odd and the behavior should actually be "strip map keys with default or null values". So arrays would remain untouched. I think that is what would actually happen for repeated protobuf values. I will update this code to only remove default/null map keys. How does that solution sound?